cv32e40s/tests/programs/custom/debug_test2/debug_test2.c

```diff
@@ -0,0 +1,3814 @@
+    sw tp, 32(sp)
+    sw t0, 36(sp)
+    sw t1, 40(sp)
+    sw t2, 44(sp)
+    sw t3, 48(sp)
+    sw t4, 52(sp)
+    sw t5, 56(sp)
+    sw t6, 60(sp)
+
+    sw gp, 64(sp)
+
+    # Get test number
+    lw s1, g_debug_test_num
+    lw s1, 0(s1)
+
+    addi s0, zero, 11
+    bne s0, s1, 1f
+    lw s0, g_debug_status
+    lw s1, 0(s0)
+    addi s1, zero, 1
+    sw s1, 0(s0)
+
+    1:
+
+    ## restore stack
+    lw gp, 64(sp)
+
+    # Restore temporary registers
+    lw t6, 60(sp)
+    lw t5, 56(sp)
+    lw t4, 52(sp)
+    lw t3, 48(sp)
+    lw t2, 44(sp)
+    lw t1, 40(sp)
+    lw t0, 36(sp)
+    lw tp, 32(sp)
+
+    # Restore argument registers
+    lw a7, 28(sp)
+    lw a6, 24(sp)
+    lw a5, 20(sp)
+    lw a4, 16(sp)
+    lw a3, 12(sp)
+    lw a2, 8(sp)
+    lw a1, 4(sp)
+    lw a0, 0(sp)
+
+    # Restore stack ptr
+    addi sp, sp, 12
+    cm.pop {ra, s0-s11}, 112
+
+    mret
+    )");
+}
+
+// ---------------------------------------------------------------
+
+__attribute__((naked)) void u_sw_irq_handler(void) {
+
+  __asm__ volatile ( R"(
+    # addi sp, sp, -124
+
+    # Push saved regs and allocate space for the remaining 16 regs
+    cm.push {ra, s0-s11}, -112
+    addi sp, sp, -12
+
+    # Save argument registers to stack
+    # as we want to be able to call C-functions
+    # from debug
+    sw a0, 0(sp)
+    sw a1, 4(sp)
+    sw a2, 8(sp)
+    sw a3, 12(sp)
+    sw a4, 16(sp)
+    sw a5, 20(sp)
+    sw a6, 24(sp)
+    sw a7, 28(sp)
+
+    # Back up remaining temporaries
+    sw tp, 32(sp)
+    sw t0, 36(sp)
+    sw t1, 40(sp)
+    sw t2, 44(sp)
+    sw t3, 48(sp)
+    sw t4, 52(sp)
+    sw t5, 56(sp)
+    sw t6, 60(sp)
+
+    sw gp, 64(sp)
+
+    # Get test number
+    lw s1, g_debug_test_num
+    lw s1, 0(s1)
+
+    #lw a0, entered_exc_handler_msg
+    #call printf
+
+    # Jump to test specific code
+    addi s0, zero, 8
+    beq s0, s1, 8f
+    addi s0, zero, 9
+    beq s0, s1, 9f
+    addi s0, zero, 11
+    beq s0, s1, 11f
+
+    beq zero, zero, 98f
+
+    8:
+    c.ebreak
+    beq zero, zero, 99f
+
+    9:
+    lw s1, g_debug_status
+    lw s0, 0(s1)
+    addi s0, zero, 1
+    sw s0, 0(s1)
+    beq zero, zero, 99f
+
+    11:
+    call check_irq_handler
+    beq zero, zero, 99f
+
+    # FIXME name?
+    98:
+    call u_sw_irq_handler_default
+    beq zero, zero, 99f # Added to prevent omission if adding further cases
+
+    99:
+    # increment mepc by appropriate amount
+    add a0, zero, zero
+    call increment_mepc
+    # old code
+    # csrrw a0, mepc, zero
+    # add a1, a0, zero
+    # call mepc_get_increment
+    # add a0, a1, a0
+    # csrrw zero, mepc, a0
+
+    # set g_handler_triggered
+    lw s0, g_handler_triggered
+    lw s1, 0(s0)
+    addi s1, zero, 1
+    sw s1, 0(s0)
+
+    # clear g_expect_ebreak
+    lw s0, g_expect_ebreak
+    lw s1, 0(s0)
+    addi s1, zero, 0
+    sw s1, 0(s0)
+
+    # clear g_expect_illegal
+    lw s0, g_expect_illegal
+    lw s1, 0(s0)
+    addi s1, zero, 0
```

Comment:
This makes the above `lw s1, 0(s0)` superfluous?

Reply:
Well spotted - No need to read the value when clearing